File: agent/agent.py
# DineAI Voice Agent

import os
import aiohttp
from dotenv import load_dotenv
from livekit import agents
from livekit.agents import AgentServer, AgentSession, Agent
from livekit.agents.llm import function_tool
from livekit.plugins import openai, deepgram, silero

# Load secrets from .env.local
load_dotenv(dotenv_path=".env.local")

BACKEND_URL = os.getenv("BACKEND_API_URL", "http://localhost:3001")
GROQ_API_KEY = os.getenv("GROQ_API_KEY")


# Tool 1: Get Weather
@function_tool()
async def get_weather(date: str, location: str = "Mumbai") -> str:
    """Fetch weather for booking date
    
    Args:
        date: Must be in YYYY-MM-DD format (e.g., "2026-02-15")
        location: Actual city name (e.g., "Mumbai", "Delhi") - DO NOT use phrases like "your location"
    """
    async with aiohttp.ClientSession() as session:
        async with session.get(
            f"{BACKEND_URL}/api/weather",
            params={"date": date, "location": location}
        ) as resp:
            data = await resp.json()
            return str(data.get("data", {}))


# Tool 2: Create Booking
@function_tool()
async def create_booking(
    customer_name: str,
    number_of_guests: int,
    booking_date: str,
    booking_time: str,
    cuisine_preference: str,
    seating_preference: str = "indoor",
    special_requests: str = ""
) -> str:
    """Save booking to database"""
    async with aiohttp.ClientSession() as session:
        async with session.post(
            f"{BACKEND_URL}/api/bookings",
            json={
                "customerName": customer_name,
                "numberOfGuests": number_of_guests,
                "bookingDate": booking_date,
                "bookingTime": booking_time,
                "cuisinePreference": cuisine_preference,
                "seatingPreference": seating_preference,
                "specialRequests": special_requests
            }
        ) as resp:
            data = await resp.json()
            booking_id = data.get("data", {}).get("bookingId", "Unknown")
            return f"Booking confirmed! ID: {booking_id}"


# The AI Assistant
class DineAIAssistant(Agent):
    def __init__(self):
        super().__init__(
            instructions="""You are DineAI, a restaurant booking assistant.
            
            FLOW:
            1. Ask name
            2. Ask number of guests
            3. Ask date and time
            4. Call get_weather function, suggest seating based on result
            5. Ask cuisine preference
            6. Ask special requests
            7. Confirm all details with user
            8. MUST call create_booking function to save - you will get booking ID from it
            
            CRITICAL RULES:
            - NEVER make up a booking ID - you MUST call create_booking to get one
            - Ask ONE question at a time
            - Be brief and friendly""",
            tools=[get_weather, create_booking]
        )


# Server Setup
server = AgentServer()


@server.rtc_session()
async def my_agent(ctx: agents.JobContext):
    """Main entry point when user connects"""
    session = AgentSession(
        stt=deepgram.STT(),           # Ears - Listens to user
        llm=openai.LLM(               # Brain - Thinks and responds
            base_url="https://api.groq.com/openai/v1",
            api_key=GROQ_API_KEY,
            model="llama-3.3-70b-versatile"
        ),
        tts=deepgram.TTS(),           # Mouth - Speaks to user
        vad=silero.VAD.load()         # Detects when user is speaking
    )
    
    await session.start(room=ctx.room, agent=DineAIAssistant())
    await session.generate_reply(instructions="Greet the user and ask for their name.")


# Start the agent
if __name__ == "__main__":
    agents.cli.run_app(server)

File: agent/agent.ts
/**
 * DineAI Voice Agent - Main Entry Point (Node.js)
 * 
 * What this file does:
 * - Creates a LiveKit voice agent for restaurant booking
 * - Connects to LiveKit Cloud to receive voice input
 * - Uses OpenAI Realtime API for voice conversation
 * - Handles the booking conversation flow
 * 
 * How to run:
 * - Development: npm run dev (connects to LiveKit Cloud)
 * - Use LiveKit Playground to test: https://agents-playground.livekit.io/
 */

import { config } from 'dotenv';

// Load environment variables from .env.local
config({ path: '.env.local' });

import { defineAgent, multimodal } from '@livekit/agents';
import { openai } from '@livekit/agents-plugin-openai';
import axios from 'axios';

// Backend API URL
const BACKEND_URL = process.env.BACKEND_API_URL || 'http://localhost:3001';

/**
 * System instructions for the AI agent
 */
const SYSTEM_INSTRUCTIONS = `You are DineAI, a friendly and professional voice assistant that helps customers book restaurant tables over voice conversation.

## Your Personality
- Warm, welcoming, and professional like a top restaurant host
- Speak naturally and conversationally, not like a robot
- Use simple words and short sentences (this is voice, not text!)
- Be enthusiastic about helping customers have a great dining experience

## Your Main Goal
Help customers book a restaurant table by collecting the following information IN THIS ORDER:
1. Customer name
2. Number of guests
3. Preferred date and time
4. Cuisine preference (Italian, Chinese, Indian, Mexican, Japanese, Thai, or American)
5. Special requests (birthday, anniversary, dietary restrictions, etc.)

## Important Conversation Rules
- Ask for ONE piece of information at a time
- Always confirm what you heard before moving on
- Keep responses SHORT - aim for 1-2 sentences when possible
- Use natural transitions like "Great!", "Perfect!", "Sounds wonderful!"

## Weather Integration
- When the user provides a booking date, fetch weather information
- Based on the weather, suggest indoor or outdoor seating

## Booking Flow Example
1. Greet: "Welcome to DineAI! I'd be happy to help you book a table. What's your name?"
2. Name: "Nice to meet you! How many guests will be joining?"
3. Guests: "Perfect, a table for 4! What date would you like to come in?"
4. Date/Time: "Wonderful! And what time works best for you?"
5. Weather Check: "The weather looks lovely that day! Would you prefer indoor or outdoor seating?"
6. Cuisine: "Great choice! What type of cuisine are you in the mood for?"
7. Special Requests: "Is this for any special occasion?"
8. Confirm: "Let me confirm and create your booking..."

Remember: You are having a VOICE conversation. Keep responses concise and natural!`;

/**
 * Helper function to fetch weather
 */
async function getWeather(date: string, location: string = 'Mumbai'): Promise<any> {
    try {
        const response = await axios.get(`${BACKEND_URL}/api/weather`, {
            params: { date, location }
        });
        return response.data.data;
    } catch (error) {
        console.error('Weather fetch error:', error);
        return null;
    }
}

/**
 * Helper function to create booking
 */
async function createBooking(bookingData: any): Promise<any> {
    try {
        const response = await axios.post(`${BACKEND_URL}/api/bookings`, bookingData);
        return response.data;
    } catch (error) {
        console.error('Booking creation error:', error);
        return null;
    }
}

/**
 * Define the DineAI Voice Agent
 */
export default defineAgent({
    entry: async (ctx) => {
        console.log('');
        console.log('🍽️ ========================================');
        console.log('🍽️ DineAI Voice Agent - New Session');
        console.log('🍽️ ========================================');
        console.log(`📍 Room: ${ctx.room.name}`);
        console.log('');

        // Connect to the room
        await ctx.connect();

        // Create OpenAI Realtime session
        const model = new openai.realtime.RealtimeModel({
            instructions: SYSTEM_INSTRUCTIONS,
            voice: 'alloy', // Natural voice
            temperature: 0.7,
            modalities: ['text', 'audio'],
        });

        // Start multimodal agent session
        const agent = new multimodal.MultimodalAgent({ model });

        // Start the session
        const session = await agent
            .start(ctx.room)
            .then(() => {
                console.log('✅ Session started - Agent is listening');

                // Send initial greeting
                session.conversation.item.create({
                    type: 'message',
                    role: 'assistant',
                    content: [
                        {
                            type: 'input_text',
                            text: "Welcome to DineAI! I'd be happy to help you book a table. What's your name?"
                        }
                    ]
                });

                session.response.create();
            });

        console.log('🎤 Greeting sent - conversation started!');
    },
});
